src/compare.c

```diff
@@ -1,136 +1,325 @@
-        } else {
-          return dcmp(xi, yj);
-        }
+        return -1;
+      }
+    } else if (R_IsNaN(xi)) {
+      if (R_IsNaN(yj)) {
+        return 0;
+      } else if (R_IsNA(yj)) {
+        return -1;
+      } else {
+        return -1;
       }
     } else {
-      return (isnan(xi) || isnan(yj)) ? NA_INTEGER : dcmp(xi, yj);
-    }
-  }
-  case STRSXP: {
-    SEXP xi = STRING_ELT(x, i), yj = STRING_ELT(y, j);
-    if (na_equal) {
-      if (xi == NA_STRING) {
-        return (yj == NA_STRING) ? 0 : -1;
+      if (R_IsNaN(yj)) {
+        return 1L;
+      } else if (R_IsNA(yj)) {
+        return 1L;
       } else {
-        return (yj == NA_STRING) ? 1 : scmp(xi, yj);
+        return dcmp(xi, yj);
       }
+    }
+  } else {
+    return (isnan(xi) || isnan(yj)) ? NA_INTEGER : dcmp(xi, yj);
+  }
+}
+
+static int chr_compare_scalar(const SEXP* x, const SEXP* y, bool na_equal) {
+  const SEXP xi = *x;
+  const SEXP yj = *y;
+
+  if (na_equal) {
+    if (xi == NA_STRING) {
+      return (yj == NA_STRING) ? 0 : -1;
     } else {
-      return (xi == NA_STRING || yj == NA_STRING) ? NA_INTEGER : scmp(xi, yj);
+      return (yj == NA_STRING) ? 1 : scmp(xi, yj);
     }
+  } else {
+    return (xi == NA_STRING || yj == NA_STRING) ? NA_INTEGER : scmp(xi, yj);
+  }
+}
 
+static int df_compare_scalar(SEXP x, R_len_t i, SEXP y, R_len_t j, bool na_equal, int n_col) {
+  int cmp;
+
+  for (int k = 0; k < n_col; ++k) {
+    SEXP col_x = VECTOR_ELT(x, k);
+    SEXP col_y = VECTOR_ELT(y, k);
+
+    cmp = compare_scalar(col_x, i, col_y, j, na_equal);
+
+    if (cmp != 0) {
+      return cmp;
+    }
   }
-  case VECSXP:
-    if (is_data_frame(x)) {
-      int p = Rf_length(x);
-      if (p != Rf_length(y))
-        stop_not_comparable(x, y, "different number of columns");
-      if (!equal_names(x, y))
-        stop_not_comparable(x, y, "different column names");
 
-      if (p == 0)
-        stop_not_comparable(x, y, "data frame with zero columns");
+  return cmp;
+}
+
+// -----------------------------------------------------------------------------
 
-      if (p > 1) {
-        for (int k = 0; k < (p - 1); ++k) {
-          SEXP col_x = VECTOR_ELT(x, k);
-          SEXP col_y = VECTOR_ELT(y, k);
-          int cmp = compare_scalar(col_x, i, col_y, j, na_equal);
+// [[ include("vctrs.h") ]]
+int compare_scalar(SEXP x, R_len_t i, SEXP y, R_len_t j, bool na_equal) {
+  switch (TYPEOF(x)) {
+  case LGLSXP: return lgl_compare_scalar(LOGICAL(x) + i, LOGICAL(y) + j, na_equal);
+  case INTSXP: return int_compare_scalar(INTEGER(x) + i, INTEGER(y) + j, na_equal);
+  case REALSXP: return dbl_compare_scalar(REAL(x) + i, REAL(y) + j, na_equal);
+  case STRSXP: return chr_compare_scalar(STRING_PTR(x) + i, STRING_PTR(y) + j, na_equal);
+  default: break;
+  }
 
-          if (cmp != 0)
-            return cmp;
-        }
-      }
+  switch (vec_proxy_typeof(x)) {
+  case vctrs_type_list: stop_not_comparable(x, y, "lists are not comparable");
```

Comment:
It seems weird that we can compare list-columns but not lists, like an invariant is broken. Perhaps we should be consistent.

Reply:
I was _very_ surprised that we could do a `vec_compare()` on a data frame with list columns. This test was meant to be one saying "cannot compare data frames with list columns" until I realized it worked. 

https://github.com/r-lib/vctrs/pull/649/files#diff-190e0bcf1adfd9d01b92248770e25c0cR85

We don't actually do a true comparison. The list column gets replaced with `vec_seq_along()` in both `x` and `y`, so essentially it gets skipped.

This happens in `vec_proxy_compare.data.frame()` where we set `vec_proxy_compare(relax = TRUE)`.

Reply:
This makes sense when you consider `arrange()` on a data frame containing list columns. We can't have an error there, and a descent into list elements would be costly and would not produce meaningful ordering.